### DISCUSSION TITLE
Parallelization with Rayon `par_iter` from query results

### DISCUSSION THREAD

**Discussion by nyurik** [rusqlite/rusqlite#1510] [Questions]
I need to do CPU-heavy parallel processing on the query results.  [Rayon](https://docs.rs/rayon/latest/rayon/iter/) is mostly for `iter()`, but it seems Rusqlite `conn.prepare(&sql).query([])` appears to be an iterator, but does not implement some needed traits, so cannot be used with `par_iter()`.  Is there a similar way to do this?

**Comment by gwenn** [comment]
Maybe with multiple connections:
https://sqlite.org/forum/info/e10528690699ee23
if the query is heavy.
Or send results to a channel with multiple consumers.

**Reply by nyurik** [reply]
@gwenn thx, but is there a reason rusqlite does not implement a standard iterator for query results?  Rayon would then be able to handle the parallelization aspect of a regular iterator.  Or am I missing something? Thx!

**Comment by gwenn** [comment]
https://docs.rs/rusqlite/latest/rusqlite/struct.Rows.html#method.next
> This interface is not compatible with Rust’s Iterator trait, because the lifetime of the returned row is tied to the lifetime of self. This is a fallible “streaming iterator”. For a more natural interface, consider using [query_map](https://docs.rs/rusqlite/latest/rusqlite/struct.Statement.html#method.query_map) or [query_and_then](https://docs.rs/rusqlite/latest/rusqlite/struct.Statement.html#method.query_and_then) instead, which return types that implement Iterator.

**Reply by nyurik** [reply]
I tried `query_map` as well, but get the same compilation error.  I was thinking of this type of pattern:

```
use rayon::prelude::*;

conn.prepare("SELECT name FROM people")
    .unwrap()
    .query_map([], |row| row.get::<_, String>(0))
    .unwrap()
    .into_par_iter()
    .for_each(|x| println!("{:?}", x));
```

```
error[E0599]: the method `into_par_iter` exists for struct `MappedRows<'_, {closure@copier.rs:569:28}>`, but its trait bounds were not satisfied
   --> mbtiles/src/copier.rs:571:14
    |
567 | /         conn.prepare("SELECT name FROM people")
568 | |             .unwrap()
569 | |             .query_map([], |row| row.get::<_, String>(0))
570 | |             .unwrap()
571 | |             .into_par_iter()
    | |_____________-^^^^^^^^^^^^^
    |
   ::: <snip>/rusqlite-0.30.0/src/row.rs:143:1
    |
143 |   pub struct MappedRows<'stmt, F> {
    |   ------------------------------- doesn't satisfy `_: IntoParallelIterator` or `_: ParallelIterator`
    |
    = note: the following trait bounds were not satisfied:
            `MappedRows<'_, {closure@mbtiles/src/copier.rs:569:28: 569:33}>: rayon::iter::ParallelIterator`
            which is required by `MappedRows<'_, {closure@mbtiles/src/copier.rs:569:28: 569:33}>: rayon::iter::IntoParallelIterator`
            `&MappedRows<'_, {closure@mbtiles/src/copier.rs:569:28: 569:33}>: rayon::iter::ParallelIterator`
            which is required by `&MappedRows<'_, {closure@mbtiles/src/copier.rs:569:28: 569:33}>: rayon::iter::IntoParallelIterator`
            `&mut MappedRows<'_, {closure@mbtiles/src/copier.rs:569:28: 569:33}>: rayon::iter::ParallelIterator`
            which is required by `&mut MappedRows<'_, {closure@mbtiles/src/copier.rs:569:28: 569:33}>: rayon::iter::IntoParallelIterator`

```

**Reply by gwenn** [reply]
https://users.rust-lang.org/t/using-rayon-with-trait-objects/52223 ?
(https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=de7cf47891036c6d4a1ba513d68e8f81)

**Reply by nyurik** [reply]
@gwenn thx, but I must still be missing something - your link doesn't point to a compiling code, nor can I figure out how it applies to my question :(

**Reply by gwenn** [reply]
You didn't provide an example to reproduce so I had to.
And you didn't read the post:
> For Vec<T>, into_par_iter() and par_iter_mut() will both require T: Send (respectively for T itself or &mut T: Send), whereas par_iter() requires T: Sync to satisfy &T: Send. You are correct to use dyn Trait + Send for the former, or you might want + Sync for the latter, or both!

vs
https://docs.rs/rusqlite/latest/rusqlite/struct.MappedRows.html
